src/Icicle/Avalanche/Simp.hs

```diff
@@ -17,22 +17,22 @@ import              Icicle.Avalanche.Program
 
 import              P
 
-simpAvalanche :: (Show n, Show p, Ord n, Eq p) => Program n p -> Fresh n (Program n p)
-simpAvalanche p
- = do p' <- transformX return simp p
-      s' <- (forwardStmts $ pullLets $ statements p')
-         >>= thresher
-         >>= forwardStmts
-         >>= nestBlocks
-         >>= thresher
+simpAvalanche :: (Show n, Show p, Ord n, Eq p) => a -> Program a n p -> Fresh n (Program a n p)
+simpAvalanche a_fresh p
+ = do p' <- transformX return (simp a_fresh) p
+      s' <- (forwardStmts a_fresh $ pullLets $ statements p')
+         >>= thresher     a_fresh
+         >>= forwardStmts a_fresh
+         >>= nestBlocks   a_fresh
+         >>= thresher     a_fresh
```

Comment:
I considered that for situations like this, perhaps we should stash an annotation in the state of `Fresh n` to avoid plumbing it around.